File: backend/lib/storage.ts
import { createClient, type SupabaseClient } from "@supabase/supabase-js";

export const STORAGE_BUCKET = "railio-uploads";
export const STORAGE_URL_PREFIX = "/api/uploads";

let _client: SupabaseClient | null = null;

export function getStorage(): SupabaseClient {
  if (_client) return _client;
  const url = process.env.SUPABASE_URL;
  const key = process.env.SUPABASE_SERVICE_ROLE_KEY;
  if (!url || !key) throw new Error("SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY required");
  _client = createClient(url, key, { auth: { persistSession: false } });
  return _client;
}

// Upload a buffer to the bucket and return a stable backend-relative path
// (e.g. "/api/uploads/42/123-uuid.png") that the frontend resolves via apiUrl.
// The /api/uploads/[...key] route signs and 302-redirects on read.
export async function uploadToBucket(
  storageKey: string,
  body: Buffer | Uint8Array,
  contentType: string
): Promise<string> {
  const { error } = await getStorage()
    .storage.from(STORAGE_BUCKET)
    .upload(storageKey, body, { contentType, upsert: true });
  if (error) throw new Error(`storage upload failed: ${error.message}`);
  return `${STORAGE_URL_PREFIX}/${storageKey}`;
}

export async function signStorageKey(storageKey: string, expiresIn = 60 * 60): Promise<string> {
  const { data, error } = await getStorage()
    .storage.from(STORAGE_BUCKET)
    .createSignedUrl(storageKey, expiresIn);
  if (error || !data) throw new Error(`storage sign failed: ${error?.message}`);
  return data.signedUrl;
}

File: backend/railio/storage.py
"""Supabase Storage helpers."""

from __future__ import annotations

from functools import lru_cache
from typing import Optional

from supabase import Client, create_client

from .config import get_settings

STORAGE_BUCKET = "railio-uploads"
STORAGE_URL_PREFIX = "/api/uploads"


@lru_cache(maxsize=1)
def get_storage() -> Client:
    s = get_settings()
    if not s.supabase_url or not s.supabase_service_role_key:
        raise RuntimeError("SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY required")
    return create_client(s.supabase_url, s.supabase_service_role_key)


def upload_to_bucket(storage_key: str, body: bytes, content_type: str) -> str:
    """Upload bytes and return the backend-relative URL the frontend can fetch."""
    res = get_storage().storage.from_(STORAGE_BUCKET).upload(
        path=storage_key,
        file=body,
        file_options={"content-type": content_type, "upsert": "true"},
    )
    # supabase-py returns an UploadResponse-ish object; treat absence of exception as success.
    _ = res
    return f"{STORAGE_URL_PREFIX}/{storage_key}"


def sign_storage_key(storage_key: str, expires_in: int = 60 * 60) -> str:
    res = get_storage().storage.from_(STORAGE_BUCKET).create_signed_url(storage_key, expires_in)
    # supabase-py returns {"signedURL": "..."} or {"signedUrl": "..."} depending on version.
    url: Optional[str] = res.get("signedURL") or res.get("signedUrl")  # type: ignore[union-attr]
    if not url:
        raise RuntimeError(f"storage sign failed: {res}")
    return url


def download_bytes(storage_key: str) -> Optional[bytes]:
    try:
        data = get_storage().storage.from_(STORAGE_BUCKET).download(storage_key)
        return bytes(data) if data else None
    except Exception:
        return None
